src/utils.rs

```diff
@@ -57,14 +59,49 @@ pub fn increment_le(x: &mut [u8]) {
 pub fn add_le(x: &mut [u8], y: &[u8]) -> Result<(), ()> {
     if x.len() == y.len() {
         unsafe {
-            ffi::sodium_add(x.as_mut_ptr(), y.as_ptr(), x.len());
+                ffi::sodium_add(x.as_mut_ptr(), y.as_ptr(), x.len());
         }
         Ok(())
     } else  {
         Err(())
     }
 }
 
+/// `bin2hex()` takes raw bytes as `bin`, converts it to hexadecimal and puts the result in `hex`.
+/// `hex` contains a nul byte (\0) terminator
+/// `bin2hex()` will return Err<()> if length of `hex` is not sufficient to hold the hex representation of `bin`.
+/// Refer docs at https://download.libsodium.org/doc/helpers#hexadecimal-encoding-decoding
+pub fn bin2hex(hex: &mut [u8], bin: &[u8]) -> Result<(), ()> {
+    if hex.len() <= 2*bin.len() {
+        return Err(());
+    }
+    unsafe {
+        ffi::sodium_bin2hex(hex.as_mut_ptr() as *mut _, 2*bin.len()+1, bin.as_ptr(), bin.len());
+    }
+    Ok(())
+}
+
+/// `hex2bin()` takes bytes of a hexadecimal string in `hex`, converts it to raw bytes and puts the result in `bin`
+/// `ignore` is an optional parameter, taking a byte representation of a string to skip
+/// `bin_len` is the length of raw bytes that `hex` was converted to.
+/// `hex_end` is an optional parameter, if passed a pointer, it will be set to the address of the
```

Comment:
So memory addresses are commonly represented in hexadecimal themselves. Does this function fill `hex_end` with something like an encoding of "0x7f06092ac6d0" and is that really the most useful thing to us? I feel like in Rust we'd want the index of the nonvalid character or a reference to it like `&hex[8]`.

If we're going to expose this field we should ensure we know what it's doing and ensure that it provides a useful utility to a Rust programmer (potentially changing the API as above), or otherwise hardcode it to null. And if we don't hardcode it to null, we need to add tests for it as well.

Reply:
Will try to make it similar to libsodium

Reply:
Updated a test to show that `hex_end` works. I like the idea of making it a reference to it like `&[u8]` but don't know how to do it. Any pointer?